### ISSUE TITLE
Clarification Needed on TensorFlow Lite Version and GPU Acceleration Configuration

### ISSUE THREAD

**Issue by bglgwyng** [mrousavy/react-native-fast-tflite#83]
Thank you for developing this excellent library. I've been using the native TensorFlow Lite library directly with Swift/Kotlin and am now migrating my code to use react-native-fast-tflite. Having experience configuring TensorFlow Lite, I have a couple of questions regarding the Android configuration:

1. Version Selection: Why does this library use an older version of TensorFlow Lite (2.12.0) instead of the latest version (2.16.1)? Is there a specific reason for pinning to this older version?

2. GPU Acceleration Configuration: I noticed that 'uses-native-library' declarations are required to enable GPU acceleration. While I understand their necessity (as GPU loading fails without them), I'm curious about why they're needed here when they weren't required in my previous direct TensorFlow Lite implementations.

For context, when I previously installed TensorFlow Lite directly, the following Gradle dependencies were sufficient for GPU acceleration:

```gradle
implementation 'org.tensorflow:tensorflow-lite:2.15.0'
implementation 'org.tensorflow:tensorflow-lite-gpu:2.15.0'
implementation 'org.tensorflow:tensorflow-lite-gpu-api:2.15.0'
```

Is the need for these additional configurations related to the older TensorFlow Lite version (2.12.0) used in this library? Or am I missing something?

**Comment by mrousavy** [comment]
Hey @bglgwyng - thank you for the kind words.

Well, to be honest I don't have an answer to your questions. I built this library a while ago fairly quickly, and I didn't investigate GPU delegates any further. It worked, but I didn't use this library for any of my RN apps.
I know a lot of people use react-native-fast-tflite (and even a lot of my clients use it), but I haven't done any significant changes here since it's release.

I'd appreciate it if you find answers to those questions after just trying stuff out, maybe we can bump the version to latest, and maybe we can simplify the GPU delegate. PRs of course welcome! 

Thanks!

**Comment by mrousavy** [comment]
iirc I tried to integrate TFLite by source, to avoid having to depend on a pod or gradle input. By source was a bit too complicated because I couldn't get it to build. That would've also worked with Simulators, and is a shared C++ codebase. I haven't pursued this further, but would be cool if you could get that working

**Comment by bglgwyng** [comment]
@TkTioNG Hello! I found that you are the author of [this PR](https://github.com/mrousavy/react-native-fast-tflite/pull/77/files#diff-8fc083d75c4d06482dc1ee1c9ba3edab0192a99add6fa96679b6ed189a049c88)
Here I asked some questions related to the PR, could you please provide answers for them?

**Comment by TkTioNG** [comment]
@bglgwyng Hi!

OpenCL is not necessary compulsory to enable GPU acceleration. May be you can try to compile your app without using it?
I believe that without using OpenCL, the device will try to use the default graphics api instead ~~(I cannot guarantee you with that XD)~~

Just that some of the tflite model, and some configurations (like serialization) will need OpenCL, where the default graphics api cannot provide and will failed to be initialized.

**Comment by bglgwyng** [comment]
@TkTioNG Thanks for the brief reply! What about the version? Do you think it is ok to bump it up?

**Comment by mrousavy** [comment]
https://github.com/mrousavy/react-native-fast-tflite/pull/85

**Comment by lucksp** [comment]
@bglgwyng did you find any solution?  I also cannot load any model with GPU enable flag on iOS on models trained after 9/1/00 (approx date).

**Comment by bglgwyng** [comment]
@lucksp #85  did you try the latest version? I'm using an old model and my problem may not be the same with yours.

**Comment by lucksp** [comment]
> @lucksp #85 did you try the latest version? I'm using an old model and my problem may not be the same with yours.

I have tried many things. On iOS models trained after August 2024 fail to load on iOS if I enable CoreML delegate. So I’m a bit stuck to say the least.

**Comment by bglgwyng** [comment]
I don't know about ML file formats, but I think that you need to specify your model file than 'trained after August 2024'.

**Comment by lucksp** [comment]
> I don't know about ML file formats, but I think that you need to specify your model file than 'trained after August 2024'.

It’s a TFLite file, trained in Vertex. The model I trained and exported prior to August works with iOS delegate flag. The model trained after does not work with this library anymore when delegate flag enabled. And I have tried other trainings too with the same result. So either Google changed how they export TFLite or a bug was introduced to this package or both?  I don’t know how to debug. 
id be happy to provide a test model TFLite file.

**Comment by ippocode** [comment]
If your .tflite was produced by ai-edge-quantizer (or another modern export pipeline built on ai_edge_litert), any weight buffer whose packed data is ≥256 KB gets serialized using TFLite's external buffer-offset mechanism instead of inline data — i.e. Buffer.offset/Buffer.size point to a blob appended after the flatbuffer root table, rather than storing the bytes inline in Buffer.data. Per schema.fbs:


table Buffer {
  data:[ubyte] (force_align: 16);
  // In a model that is larger than 2GB, then buffers instead uses the following
  // attributes to find stored data, which is outside of flatbuffers
  // the offset is calculated relative to the beginning of the file and is only
  // valid if > 1.
  offset: ulong;
  size: ulong;
}
Despite that comment saying "larger than 2GB," in practice the quantizer triggers this at a much lower threshold (256 KB), so plenty of ordinary-sized models end up using it.

Why it fails on iOS specifically: the iOS TFLite runtime distributed via CocoaPods — TensorFlowLiteC — is frozen at 2.17.0 (July 2024). CocoaPods stopped receiving any TFLite/LiteRT updates after mid-2025 (checked via pod trunk info TensorFlowLiteC / LiteRTC — no stable release, no recent nightly). That old runtime doesn't resolve offset-referenced buffers, so at Invoke you get Input tensor N lacks data for whichever tensor happens to be offset-referenced (in our case, tensor 176, a conv weight). Android didn't have this problem because bumping the LiteRT Gradle dependency (1.4.0 → 1.4.2) picked up a runtime that does resolve offsets — but there's no equivalent newer pod on iOS.

Fix — lossless re-serialization, no re-export needed: round-trip the finished .tflite through LiteRT's own flatbuffer utilities to force everything back to inline storage:


from ai_edge_quantizer.utils import tfl_flatbuffer_utils

model = tfl_flatbuffer_utils.read_model("model.tflite")   # resolves every
                                                             # offset buffer
                                                             # back to inline
                                                             # .data, sets
                                                             # buffer.offset=0
tfl_flatbuffer_utils.write_model(model, "model_fixed.tflite")  # re-packs with
                                                                  # _experimental_use_buffer_offset=false
                                                                  # → zero offset
                                                                  # buffers
(Both functions just delegate to tflite.tools.flatbuffer_utils in google-ai-edge/LiteRT, so you can use that directly too if you don't want the quantizer dependency.)

This is not a re-export — the weights are bit-identical, only the on-disk buffer layout changes (external → inline). No need to touch PyTorch/the original checkpoint/quantization recipe at all — it operates on the already-exported .tflite file. Took our vision model's file size down by ~1.9 KB (just the offset-region alignment padding disappearing) and it Invokes correctly on iOS afterward.

Suggested verification if you hit this: dump model.buffers and count how many have offset > 1 (per schema, offset is only meaningful if >1) before/after the round-trip — should go from some non-zero count to 0. You can also diff outputs on desktop TF (which resolves offsets fine) before/after to confirm nothing changed numerically — should be bit-identical/cosine ≈ 1.0.
